portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -26,45 +37,102 @@
+          .build();
+    QueryResults<Entity> results = datastore.run(query);
+
+    ArrayList<UserComment> comments = new ArrayList<>();
+    while(results.hasNext()) {
+      Entity entity = results.next();
+      long id = entity.getKey().getId();
+      String name = entity.getString("name");
+      String email = entity.getString("email");
+      String time =  String.valueOf(entity.getLong("time"));
+      String comment = entity.getString("comment");
+      UserComment userComment = UserComment.create(name, email, comment, time);
+      comments.add(userComment);
+    }
+
     Gson gson = new Gson();
-    String json = gson.toJson(userComments);
     response.setContentType("application/json;");
-    response.getWriter().println(json);
+    response.getWriter().println(gson.toJson(comments));
   }
 
   /*
    * Called when a client submits a POST request to the /data URL
    * Adds submitted comment to internal record if the comment is
-   * non-empty. Then, the page is reloaded.
+   * non-empty. 
    */
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String userComment = getFieldFromResponse(request, "comment", "");
+    String parsedBody = CharStreams.toString(request.getReader());
+
+    JsonParser parser = new JsonParser();
+    JsonElement parsedJson = parser.parse(parsedBody);
+    JsonObject jsonObject = parsedJson.getAsJsonObject();
+
+    String userComment = getFieldFromJsonObject(jsonObject, "comment", "");
     if(userComment.length() != 0) {
-      String userName = getFieldFromResponse(request, "name", "Anonymous");
-      String userEmail = getFieldFromResponse(request, "email", "janedoe@gmail.com");
-      Date date = new Date();
-      String currDate = date.toString();
-      String userDate = getFieldFromResponse(request, "time", currDate);
-      UserComment comment = UserComment.create(userName, userEmail, userComment, userDate);
-      userComments.add(comment);
+      String userName = getFieldFromJsonObject(jsonObject, "name", "Anonymous");
+      String userEmail = getFieldFromJsonObject(jsonObject, "email", "janedoe@gmail.com");
+      String currDate = String.valueOf(System.currentTimeMillis());
+      long userDate = Long.parseLong(getFieldFromJsonObject(jsonObject, "timestamp", currDate));
+      addToDatastore(userName, userEmail, userDate, userComment);
     }
-    response.sendRedirect("index.html");
   }
 
   /*
-   * Extracts the value of fieldName attribute from request object if present
-   * and returns defaultValue if it is not or it is empty
+   * Extracts the value of fieldName attribute from jsonObject if present
+   * and returns defaultValue if it is not or the value is empty
    */
-    private String getFieldFromResponse(HttpServletRequest request, String fieldName, String defaultValue) {
+  public String getFieldFromJsonObject(JsonObject jsonObject, String fieldName, String defaultValue) {
```

Comment:
This could be simplified to

```
if (jsonObject.has(fieldName)) {
  String fieldValue = jsonObject.get(fieldName).getAsString();
  return fieldValue.isEmpty() ? defaultValue : fieldValue;
}
return defaultValue;
```